### ISSUE TITLE
Check ToObject() coercion semantics for Object built-ins

### ISSUE THREAD

**Issue by svaarala** [svaarala/duktape#1079]
For example, in ES5.1 `Object.getOwnPropertyNames(123)` is a TypeError but it is treated by ES6 the same as `Object.getOwnPropertyNames(Object(123))`.

Go through all Object built-ins and check for ToObject() coercion changes.

**Comment by fatcerberus** [comment]
I have a pull open for this but it kept breaking tests and I was too lazy to fix them, so... :)

**Comment by svaarala** [comment]
Hmm, do you mean a branch and not a pull?

**Comment by svaarala** [comment]
Sorry, thought this was the Linenoise pull :)

**Comment by svaarala** [comment]
Right, #1028 could fix this behavior.

**Comment by fatcerberus** [comment]
Yeah, #1028 is the pull.

**Comment by svaarala** [comment]
Fixing this is bound to break testcases as you noted in #1028 so there's some trivia fixing to do.

Further, if this is implemented without config options for the behavior, it means you can't compile an ES5 compliant build from master because that throwing behavior is strictly mandatory in ES5/ES5.1. But maintaining variable ES5 and post-ES5 behavior is also awkward extra work. I've been able to skirt this issue to a reasonable extent so far but I guess it's decision time :-)

My preference would be to just keep master tracking the latest specification for all the implemented behavior - but make it possible to configure for ES5ish build where the bindings are from ES5 but semantics trivia like coercion behavior may be from ES6 or above. This should work for all users except those wanting exact ES5/ES5.1 compliance.

If taking the route of maintaining an ES5.1 build, test coverage will be needed for ES5.1 and ES-latest and it must be possible to write testcases targeting the ES5.1 variant vs. the ES-latest variant.

**Comment by svaarala** [comment]
So to be clear, I think that ES5.1 variant issue needs to be thought out first before making the change. Otherwise there would be a de facto decision in master (= ES-latest only).

**Comment by svaarala** [comment]
There's also a grey area issue for versioning when changes are made to align better with ES6: those changes can be thought of as fixing bugs in ES6 (or ES7 or whatever) support, but they can also be viewed as incompatible changes to ES5 support.

The practical concern is that with the latter interpretation any minute change to align with ES6 requires a major release. That might mean that most releases would be major releases because ES6 features are implemented in small parts.

**Comment by fatcerberus** [comment]
I guess the question to ask is whether strict ES5 compliance is meaningful given the existence of later standards.

In the past you've described Ecmascript as a living standard.  To me that implies any changes made to the specification are meant to be retroactive - if you don't follow the current behavior then you're out of compliance.  This seems to be ECMA's intent as well, given the switch from explicit numbered versions to year-based ones ("ECMAScript 2015") as well as the fact that this page only links to the latest version (ES7 at the time of writing):
http://www.ecma-international.org/publications/standards/Ecma-262.htm

Even if one doesn't buy that "retroactive" reasoning, there's still the real-world issue that the legacy behavior will become surprising to JS developers as later specifications are adopted (case in point: https://github.com/svaarala/duktape/pull/1078#issuecomment-260234237 ;-).  As such it's questionable how often such an option would be utilized in practice.  IMO it's far more likely that those strapped for footprint will simply disable bindings to keep code size down while maintaining the ES-latest behavior.

**Comment by svaarala** [comment]
> I guess the question to ask is whether strict ES5 compliance is meaningful given the existence of later standards.

Well that's a very well defined target, and there are several reasons to want an ES5 compliant environment. It's widely supported, and there are multiple (even embedded) engines providing ES5 support so you can switch between engines relatively easily. For embedded environments it's a stable target which is good if you just want a stable scripting environment instead of constantly changing one: not everyone desires the latest and greatest :)

Duktape is currently described as an ES5/ES5.1 engine with some borrowed ES6/ES7 features. That more or less implies ES5.1 is the "dominant" specification while later features are borrowed. If master adopts a "target ES-latest" approach, this would need to be changed so that Duktape targets the latest specification (but is not a complete implementation of it, yet), but with an ES5-ish subset for low memory targets. This would be a pretty clear target too.

> To me that implies any changes made to the specification are meant to be retroactive [...]

The changes in the specification are clearly "retroactive" in that they break existing behaviors and, potentially, existing code relying on those behaviors. But that doesn't really tell you what to do with versioning: semantic versioning guarantees are, on the face of it, pretty straightforward:

```
    MAJOR version when you make incompatible API changes,
    MINOR version when you add functionality in a backwards-compatible manner, and
    PATCH version when you make backwards-compatible bug fixes.
```

They're not always easy to interpret in practice IMO. For example, if you fix a bug it's perfectly possible someone was relying on that buggy behavior by accident, so that the fix is not backwards compatible in that case and a major version bump would be necessary to fix the bug.

Similarly, what does it mean to "add functionality" in concrete terms? Adding new global object bindings, for example, is not necessarily backwards compatible if some application already used a binding of that name. C API additions are much easier to understand: new functions don't matter if you don't call them.

Changing ES5 semantics already present in master automatically means a major version bump because it's not backwards compatible. This is easy to demonstrate with any piece of code that checks for the old vs. new behavior, and there's no telling if an application might rely on that behavior intentionally, or maybe more likely, by accident.

One can sort of fudge that situation by targeting "ES-latest" and treating any deviances as bugs. This means that if you install a minor upgrade, *anything* may change if a later specification version requires it. It's not very clean but it is practical.

Because semantic versioning is about managing expectations of changes in library versions, it's also important to state clearly what the functional target is. If the target is (a subset of) ES-latest, then it's not so surprising if things change to align better with later specifications.

It'd also be possible to just use semantic versioning conservatively and just make any release major if it changes any existing behaviors. That would, in practice, mean most releases would be major. But since we're not going to run out of numbers, that would be a viable approach too.

**Comment by fatcerberus** [comment]
That retroactive argument does seem to hold some merit: ECMA describes earlier specifications as "historical", having been "replaced" by later editions:

> The previous replaced "historical" editions of this Ecma Standard are available here [link].

**Comment by svaarala** [comment]
Right, but where does semantic versioning allow that to mean incompatible changes can be made in minor versions?

It's about versioning expectations, and incompatible changes in minor versions are not expected.

**Comment by svaarala** [comment]
This is from semver FAQ:

>If even the tiniest backwards incompatible changes to the public API require a major version bump, won’t I end up at version 42.0.0 very rapidly?
>
> This is a question of responsible development and foresight. Incompatible changes should not be introduced lightly to software that has a lot of dependent code. The cost that must be incurred to upgrade can be significant. Having to bump major versions to release incompatible changes means you’ll think through the impact of your changes, and evaluate the cost/benefit ratio involved.

So it's considered a feature that incompatible changes require a major version number so you'd think about the changes before making them.

Unfortunately that doesn't work very well for tracking a "living standard" because quite a lot of changes will become major if one takes that literally. It's difficult to implement features piece by piece and releasing often.

I've tried to manage this problem earlier by describing which parts are under semver guarantees and which parts are not: http://duktape.org/guide.html#versioning. This works reasonably well for a static specification target which affects the C API and the Ecmascript bindings so that e.g. `Object.getPrototypeOf()` doesn't have a yearly changing behavior ;-) For example config options, debugger protocol, and bytecode dump/load format are not under semver guarantees because that would again make most changes major.

Regardless of the outcome for this issues, that guide section should describe what semver guarantees are in place, in order to provide clear expectations of what a patch version, a minor version, and a major version mean for Duktape. Semver is, after all, about expectations when seeing a library version be updated.

**Comment by fatcerberus** [comment]
No, I get that, the argument I was making was re: whether or not to keep "strict ES5 behavior" as an option, I'm not disagreeing with the semantic versioning logic at all.  It seems we may have shifted gears somewhere along the way without me realizing it. :)

**Comment by svaarala** [comment]
Ah, I see :) This is what I've been talking about (quoting myself):

> The practical concern is that with the latter interpretation any minute change to align with ES6 requires a major release. That might mean that most releases would be major releases because ES6 features are implemented in small parts.

**Comment by svaarala** [comment]
So the grey area concern is that changing `Object.getPrototypeOf()` coercing behavior is a major release if one reads semver literally.

That seems quite odd - but we could do that.

In essence that would mean that a release would be intended initially as a minor one, but the first commit to introduce a change to existing C or Ecmascript API behavior, regardless of how small, would cause the release to be major. That makes planning milestones in advance a bit difficult but I'm sure that can be worked around.

**Comment by svaarala** [comment]
Regarding what to target, like I said above, I lean towards implementing newest features incrementally, but maintaining an ES5-like subset for low memory targets. The differences are small enough to often not matter (though I'm sure someone specifically wants ES5 compliance; I've been asked for that quite many times) and ES6+ is a bit too large for some embedded targets.

Then again, it would be equally possible to define the project target as: ES-latest, with the ability to configure subsets of it as necessary. For example, drop RegExps (ES5), but include WeakMap and symbols (ES6). There would then be no ES5 promise anymore, but instead a promise to be able to configure subsets reasonably well to control code and RAM footprint. This is already possible pretty well so maybe it would be enough as a target.

**Comment by svaarala** [comment]
Just to recap why semver entered the discussion: with a fixed ES5.1 + extras target semver is reasonably straightforward at least for the ES5.1 part. With an ES-latest target semver becomes more difficult in practice because releases tend to be major releases - or one needs to relax on the semver interpretation a bit. So that's why these two issues are somewhat connected.

But that's just about semver specifics which may need further thought if the project goal is ES-latest "living standard" rather than a more fixed target. Semver is also not fixed in stone; if it doesn't work out well then it's of course possible to change versioning schemes.

**Comment by svaarala** [comment]
Assuming the project goes with the ES-latest route and won't maintain a strict ES5 compliant configuration, the change in this issue (ToObject() coercions) should be done for 2.0 because they're backwards incompatible. The next opportunity for that would be 3.0.

**Comment by fatcerberus** [comment]
Agreed.  Presumably in that case `DUK_USE_ES6` would (eventually) be removed as well?

**Comment by svaarala** [comment]
`DUK_USE_ES6` doesn't mean "use ES6 semantics" but it's a placeholder config option for including ES6 *bindings* that don't have specific config options. It would still be needed for providing an ES5-ish environment, i.e. for footprint control.

**Comment by svaarala** [comment]
Implemented by #1028. Master is also new ES5-ish w.r.t. versioning.
